commands/template_pull_test.go

```diff
@@ -4,141 +4,123 @@ package commands
+		faasCmd.SetOutput(&buf)
+		err := faasCmd.Execute()
 
-func Test_repositoryUrlRegExp(t *testing.T) {
-	var url string
-	r := regexp.MustCompile(repositoryRegexpGithub)
+		if !strings.Contains(err.Error(), "The repository URL must be a valid git repo uri") {
+			t.Fatal("Output does not contain the required string", err.Error())
+		}
+	})
+}
 
-	url = "http://github.com/owner/repo"
-	if r.MatchString(url) {
-		t.Errorf("Url %s must start with https", url)
+func Test_repositoryUrlRemoteRegExp(t *testing.T) {
+
+	r := regexp.MustCompile(gitRemoteRepoRegex)
+	validURLs := []struct {
+		name string
+		url  string
+	}{
+		{name: "git protocol with sha", url: "git://github.com/openfaas/faas.git#ff78lf9h"},
+		{name: "git protocol with branch", url: "git://github.com/openfaas/faas.git#master"},
+		{name: "git protocol", url: "git://host.xz/path/to/repo.git/"},
+		{name: "scp style with ip address", url: "git@192.168.101.127:user/project.git"},
+		{name: "scp style with hostname", url: "git@github.com:user/project.git"},
+		{name: "http protocol with ip address", url: "http://192.168.101.127/user/project.git"},
+		{name: "http protocol", url: "http://github.com/user/project.git"},
+		{name: "https protocol with ip address", url: "https://192.168.101.127/user/project.git"},
+		{name: "https protocol with hostname", url: "https://github.com/user/project.git"},
+		{name: "https protocol with basic auth", url: "https://username:password@github.com/username/repository.git"},
+		{name: "ssh protocol with hostname no port", url: "ssh://user@host.xz/path/to/repo.git/"},
+		{name: "ssh protocol with hostname and port", url: "ssh://user@host.xz:port/path/to/repo.git/"},
 	}
 
-	url = "https://github.com/owner/repo.git"
-	if r.MatchString(url) {
-		t.Errorf("Url %s must not end with .git or must start with https", url)
-	}
+	for _, scenario := range validURLs {
+		t.Run(fmt.Sprintf("%s is a valid remote git url", scenario.name), func(t *testing.T) {
+			if !r.MatchString(scenario.url) {
+				t.Errorf("Url %s should pass the regex match", scenario.url)
+			}
 
-	url = "https://github.com/owner/repo//"
-	if r.MatchString(url) {
-		t.Errorf("Url %s must end with no or one slash", url)
+		})
 	}
 
-	url = "https://github.com/owner/repo"
-	if !r.MatchString(url) {
-		t.Errorf("Url %s must be valid", url)
+	invalidURLs := []struct {
+		name string
+		url  string
+	}{
+		{name: "local repo file protocol", url: "file:///path/to/repo.git/"},
+		{name: "ssh missing username and port", url: "host.xz:/path/to/repo.git"},
+		{name: "ssh username and missing port", url: "user@host.xz:path/to/repo.git"},
+		{name: "relative local path", url: "path/to/repo.git/"},
+		{name: "magic relative local", url: "~/path/to/repo.git"},
 	}
+	for _, scenario := range invalidURLs {
+		t.Run(fmt.Sprintf("%s is not a valid remote git url", scenario.name), func(t *testing.T) {
+			if r.MatchString(scenario.url) {
+				t.Errorf("Url %s should fail the regex match", scenario.url)
```

Comment:
Ideally print the regex in the error.

Reply:
Just added the regex to the message and rebased to fix the conflicts